File: backend/model/src/main/java/courses/paint/mini/model/course/CourseStep.java
package courses.paint.mini.model.course;

import courses.paint.mini.model.product.ModelingProduct;
import courses.paint.mini.model.product.Paint;
import lombok.*;

import java.util.Map;
import java.util.Set;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(of = {"id", "orderNumber"})
public class CourseStep {

    private String id;
    private Long orderNumber;
    private String title;
    private String description;
    private Map<Paint, PaintingTechnique> usedPaints;
    private Set<ModelingProduct> usedOtherModelingProducts;

}

File: frontend/src/shared/model/course-step.ts
export class CourseStep {

  constructor(id: string | null,
              orderNumber: number,
              title: string,
              description: string,
              paintTechniqueIdToPaintIdMap: { [name: string]: string },
              usedOtherModelingProductIds: string[]) {
    if (id) {
      this.id = id;
    }
    this.orderNumber = orderNumber;
    this.title = title;
    this.description = description;
    this.usedOtherModelingProductIds = usedOtherModelingProductIds;

    this.paintTechniqueIdToPaintIdMap = new Map;

    Object.keys(paintTechniqueIdToPaintIdMap).forEach(key => {
      this.paintTechniqueIdToPaintIdMap.set(key, paintTechniqueIdToPaintIdMap[key]);
    });
  }

  id!: string;
  orderNumber!: number;
  title!: string;
  description!: string;
  paintTechniqueIdToPaintIdMap!: Map<string, string>;
  usedOtherModelingProductIds!: string[];

}
